### PR TITLE
tests for val loop flow

### PR DESCRIPTION
**Pull request by williamFalcon**
Tests for all combinations of returns and val loop flows to ensure the eval and test loops work well

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1163,22 +1166,38 @@ def run_pretrain_routine(self, model: LightningModule):
...
-            else:
+            if len(eval_loop_results) == 0:
                 return 1
 
+            # remove the tensors from the eval results
+            for i, result in enumerate(eval_loop_results):
+                if isinstance(result, dict):
+                    for k, v in result.items():
+                        if isinstance(v, torch.Tensor):
+                            result[k] = v.cpu().item()
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1179]: shall we detouch it too?

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1197,26 +1216,20 @@ def run_pretrain_routine(self, model: LightningModule):
...
+                # when we get a list back, used only the last item
+                if isinstance(eval_results, list):
+                    eval_results = eval_results[-1]
                 _, _, _, callback_metrics, _ = self.process_output(eval_results)
                 self.callback_metrics = callback_metrics
 
             self.on_sanity_check_end()
 
-        # clear cache before training
-        if self.on_gpu and self.root_gpu is not None:
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1206]: this is not needed anymore?

File: tests/base/deterministic_model.py [hunk 3]
```diff
@@ -162,13 +166,61 @@ def training_epoch_end_dict(self, outputs):
...
         acc = self.step(batch, batch_idx)
 
-        logs = {'log_acc1': torch.tensor(12).type_as(acc), 'log_acc2': torch.tensor(7).type_as(acc)}
+        logs = {'log_acc1': torch.tensor(12 + batch_idx).type_as(acc), 'log_acc2': torch.tensor(7).type_as(acc)}
         pbar = {'pbar_acc1': torch.tensor(17).type_as(acc), 'pbar_acc2': torch.tensor(19).type_as(acc)}
         return {'val_loss': acc, 'log': logs, 'progress_bar': pbar}
 
+    def validation_step_end_no_return(self, val_step_output):
+        assert len(val_step_output) == 3
+        assert val_step_output['val_loss'] == 171
```
Comment (ananyahjha93) [inline tests/base/deterministic_model.py:193]: constants like 171 which are used to assert results in tests should be defined as constants in the deterministic_model.py.

ex- CONST_1 = 171
and then from tests.base import CONST_1
and then assert

Comment (williamFalcon) [reply tests/base/deterministic_model.py:193]: but these change a bit depending on the particular setup.
I agree that we can make them constants later, but for now i'd rather be able to iterate through the tests faster.

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py:1206]: it's still there... just moved up

**Comment by ananyahjha93** [comment]
@williamFalcon apart from the build errors (not sure why they are there), everything else looks good to me.

**Pull request merged**